physical/postgresql.go

```diff
@@ -0,0 +1,180 @@
+package physical
+
+import (
+	"database/sql"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/armon/go-metrics"
+	_ "github.com/lib/pq"
+)
+
+// PostgreSQL Backend is a physical backend that stores data
+// within a PostgreSQL database.
+type PostgreSQLBackend struct {
+	table      string
+	client     *sql.DB
+	statements map[string]*sql.Stmt
+}
+
+// newPostgreSQLBackend constructs a PostgreSQL backend using the given
+// API client, server address, credentials, and database.
+func newPostgreSQLBackend(conf map[string]string) (Backend, error) {
+	// Get the PostgreSQL credentials to perform read/write operations.
+	connURL, ok := conf["connection_url"]
+	if !ok || connURL == "" {
+		return nil, fmt.Errorf("missing connection_url")
+	}
+
+	table, ok := conf["table"]
+	if !ok {
+		table = "vault"
+	}
+
+	key_column, ok := conf["key_column"]
+	if !ok {
+		key_column = "vault_key"
+	}
+
+	value_column, ok := conf["value_column"]
+	if !ok {
+		value_column = "vault_value"
+	}
+
+	upsert_function, ok := conf["upsert_function"]
+	if !ok {
+		upsert_function = "vault_upsert"
+	}
+
+	// Create PostgreSQL handle for the database.
+	db, err := sql.Open("postgres", connURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to postgres: %v", err)
+	}
+
+	// Determine if we should use an upsert function (versions < 9.5)
+	var upsert_required bool
+	upsert_required_query := "SELECT string_to_array(setting, '.')::int[] < '{9,5}' FROM pg_settings WHERE name = 'server_version'"
+	if err := db.QueryRow(upsert_required_query).Scan(&upsert_required); err != nil {
+		return nil, fmt.Errorf("failed to check for native upsert: %v", err)
+	}
+
+	var put_statement string
+	if upsert_required {
+		put_statement = "SELECT " + upsert_function + "($1, $2)"
+	} else {
+		put_statement = "INSERT INTO " + table + " VALUES($1, $2)" +
```

Comment:
Also [make sure to quote any identifiers taken from config](https://github.com/lib/pq/blob/master/conn.go#L1481).